File: all-examples/go/v3/main.go
package main

import (
	"context"
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/aws/amazon-s3-encryption-client-go/v3/client"
	"github.com/aws/amazon-s3-encryption-client-go/v3/materials"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/kms"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func main() {
	// Check command line arguments
	if len(os.Args) != 5 {
		fmt.Printf("Usage: %s <bucket-name> <object-key> <kms-key-id> <region>\n", os.Args[0])
		fmt.Printf("Example: %s avp-21638 s3ec-go-v3 arn:aws:kms:us-east-2:648638458147:key/a47079da-17e4-45a5-b82e-2bac101cad01 us-east-2\n", os.Args[0])
		os.Exit(1)
	}

	bucketName := os.Args[1]
	objectKey := os.Args[2]
	kmsKeyID := os.Args[3]
	region := os.Args[4]

	fmt.Println("=== S3 Encryption Client v3 Example (Go) ===")
	fmt.Printf("Bucket: %s\n", bucketName)
	fmt.Printf("Object Key: %s\n", objectKey)
	fmt.Printf("KMS Key ID: %s\n", kmsKeyID)
	fmt.Printf("Region: %s\n", region)
	fmt.Println()

	// Test data for encryption
	testData := "Hello, World! This is a test message for S3 encryption client v3 in Go."
	fmt.Printf("Original data: %s\n", testData)
	fmt.Printf("Data length: %d bytes\n", len(testData))
	fmt.Println()

	fmt.Println("--- Initialize S3 Encryption Client v3 ---")

	// Create regular S3 client
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(region))
	if err != nil {
		fmt.Printf("Error loading AWS config: %v\n", err)
		os.Exit(1)
	}
	s3Client := s3.NewFromConfig(cfg)

	// Create KMS client
	kmsClient := kms.NewFromConfig(cfg)

	// Create KMS keyring
	keyring := materials.NewKmsKeyring(kmsClient, kmsKeyID)

	// Create Cryptographic Materials Manager
	cmm, err := materials.NewCryptographicMaterialsManager(keyring)
	if err != nil {
		fmt.Printf("Error creating CMM: %v\n", err)
		os.Exit(1)
	}

	// Create S3 Encryption Client v3
	encryptionClient, err := client.New(s3Client, cmm)
	if err != nil {
		fmt.Printf("Error creating S3 Encryption Client: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Successfully initialized S3 Encryption Client v3")
	fmt.Println("--- Encrypt and Upload Object to S3 ---")

	// Add encryption context for additional security
	encryptionContext := map[string]string{
		"purpose":  "example",
		"version":  "v3",
		"language": "go",
	}

	// Create context with encryption context
	ctx := context.WithValue(context.Background(), "EncryptionContext", encryptionContext)

	// Upload encrypted object using S3 Encryption Client
	putInput := &s3.PutObjectInput{
		Bucket:   aws.String(bucketName),
		Key:      aws.String(objectKey),
		Body:     strings.NewReader(testData),
		Metadata: encryptionContext,
	}

	_, err = encryptionClient.PutObject(ctx, putInput)
	if err != nil {
		if strings.Contains(err.Error(), "NoSuchBucket") {
			fmt.Printf("Error: S3 bucket '%s' does not exist or is not accessible\n", bucketName)
		} else if strings.Contains(err.Error(), "NotFoundException") {
			fmt.Printf("Error: KMS key '%s' not found or not accessible\n", kmsKeyID)
		} else {
			fmt.Printf("Error uploading encrypted object: %v\n", err)
		}
		os.Exit(1)
	}

	fmt.Println("Successfully uploaded encrypted object to S3!")
	fmt.Printf("   Bucket: %s\n", bucketName)
	fmt.Printf("   Key: %s\n", objectKey)
	fmt.Printf("   Encryption Context: %v\n", encryptionContext)
	fmt.Println()

	fmt.Println("--- Download and Decrypt Object from S3 ---")

	// Download and decrypt object using S3 Encryption Client
	getInput := &s3.GetObjectInput{
		Bucket: aws.String(bucketName),
		Key:    aws.String(objectKey),
	}

	getResponse, err := encryptionClient.GetObject(ctx, getInput)
	if err != nil {
		fmt.Printf("Error downloading and decrypting object: %v\n", err)
		os.Exit(1)
	}
	defer getResponse.Body.Close()

	// Read the decrypted data
	decryptedData, err := io.ReadAll(getResponse.Body)
	if err != nil {
		fmt.Printf("Error reading decrypted data: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Successfully downloaded and decrypted object from S3!")
	fmt.Printf("   Object size: %d bytes\n", len(decryptedData))
	fmt.Printf("   Decrypted data: %s\n", string(decryptedData))
	fmt.Println()

	fmt.Println("--- Verify Roundtrip Success ---")

	// Verify the roundtrip was successful
	if string(decryptedData) == testData {
		fmt.Println("SUCCESS: Roundtrip encryption/decryption completed successfully!")
		fmt.Println("   Original data matches decrypted data")
		fmt.Println("   Data integrity verified")
	} else {
		fmt.Println("ERROR: Roundtrip failed - data mismatch")
		fmt.Printf("   Original: %s\n", testData)
		fmt.Printf("   Decrypted: %s\n", string(decryptedData))
		os.Exit(1)
	}

	// Optionally Delete the Object
	//fmt.Println("--- Cleanup ---")
	// Clean up the test object using regular S3 client
	// _, err = s3Client.DeleteObject(context.TODO(), &s3.DeleteObjectInput{
	// 	Bucket: aws.String(bucketName),
	// 	Key:    aws.String(objectKey),
	// })
	// if err != nil {
	// 	fmt.Printf("Error deleting test object: %v\n", err)
	// } else {
	// 	fmt.Println("Test object deleted from S3")
	// }

	fmt.Println()
	fmt.Println("=== Example completed successfully! ===")
}

File: all-examples/ruby/v3/main.rb
#!/usr/bin/env ruby

require 'aws-sdk-s3'
require 'aws-sdk-kms'
require 'json'

# See: https://github.com/ruby/openssl/issues/949
Aws.use_bundled_cert!

def main
  # Check command line arguments
  if ARGV.length != 4
    puts "Usage: #{$0} <bucket-name> <object-key> <kms-key-id> <region>"
    puts "Example: #{$0} avp-21638 s3ec-ruby-v3 arn:aws:kms:us-east-2:648638458147:key/a47079da-17e4-45a5-b82e-2bac101cad01 us-east-2"
    exit 1
  end

  bucket_name = ARGV[0]
  object_key = ARGV[1]
  kms_key_id = ARGV[2]
  region = ARGV[3]

  puts "=== S3 Encryption Client v3 Example (Ruby) ==="
  puts "Bucket: #{bucket_name}"
  puts "Object Key: #{object_key}"
  puts "KMS Key ID: #{kms_key_id}"
  puts "Region: #{region}"
  puts

  begin
    # Test data for encryption
    test_data = "Hello, World! This is a test message for S3 encryption client v3 in Ruby."
    puts "Original data: #{test_data}"
    puts "Data length: #{test_data.length} bytes"
    puts

    puts "--- Initialize S3 Encryption Client v3 ---"
    
    # Create regular S3 client
    s3_client = Aws::S3::Client.new(region: region)
    
    # Create KMS client
    kms_client = Aws::KMS::Client.new(region: region)
    
    # Create S3 Encryption Client v3
    encryption_client = Aws::S3::EncryptionV3::Client.new(
      client: s3_client,
      kms_key_id: kms_key_id,
      kms_client: kms_client,
      key_wrap_schema: :kms_context
    )
    
    puts "Successfully initialized S3 Encryption Client v3"
    puts "--- Encrypt and Upload Object to S3 ---"
    
    # Add encryption context
    encryption_context = {
      'purpose' => 'example',
      'version' => 'v3',
      'language' => 'ruby'
    }
    
    # Upload encrypted object using S3 Encryption Client
    put_response = encryption_client.put_object({
      bucket: bucket_name,
      key: object_key,
      body: test_data,
      kms_encryption_context: encryption_context
    })
    
    puts "Successfully uploaded encrypted object to S3!"
    puts "   Bucket: #{bucket_name}"
    puts "   Key: #{object_key}"
    puts "   Encryption Context: #{encryption_context}"
    puts

    puts "--- Download and Decrypt Object from S3 ---"
    
    # Download and decrypt object using S3 Encryption Client
    get_response = encryption_client.get_object({
      bucket: bucket_name,
      key: object_key,
      kms_encryption_context: encryption_context
    })
    
    # Read the decrypted data
    decrypted_data = get_response.body.read
    
    puts "Successfully downloaded and decrypted object from S3!"
    puts "   Object size: #{decrypted_data.length} bytes"
    puts "   Decrypted data: #{decrypted_data}"
    puts

    puts "--- Verify Roundtrip Success ---"
    
    # Verify the roundtrip was successful
    if decrypted_data == test_data
      puts "SUCCESS: Roundtrip encryption/decryption completed successfully!"
      puts "   Original data matches decrypted data"
      puts "   Data integrity verified"
    else
      puts "ERROR: Roundtrip failed - data mismatch"
      puts "   Original: #{test_data}"
      puts "   Decrypted: #{decrypted_data}"
      exit 1
    end

    # Optionally Delete the Object
    #puts "--- Cleanup ---"
    # Clean up the test object using regular S3 client
    # s3_client.delete_object({
    #   bucket: bucket_name,
    #   key: object_key
    # })
    # puts "Test object deleted from S3"
    
    puts
    puts "=== Example completed successfully! ==="

  rescue Aws::S3::Errors::NoSuchBucket => e
    puts "Error: S3 bucket '#{bucket_name}' does not exist or is not accessible"
    puts "   #{e.message}"
    exit 1
  rescue Aws::KMS::Errors::NotFoundException => e
    puts "Error: KMS key '#{kms_key_id}' not found or not accessible"
    puts "   #{e.message}"
    exit 1
  rescue Aws::S3::EncryptionV3::Errors::EncryptionError => e
    puts "S3 Encryption Error: #{e.message}"
    exit 1
  rescue Aws::S3::EncryptionV3::Errors::DecryptionError => e
    puts "S3 Decryption Error: #{e.message}"
    exit 1
  rescue Aws::Errors::ServiceError => e
    puts "AWS Service Error: #{e.message}"
    puts "   Error Code: #{e.code}" if e.respond_to?(:code)
    exit 1
  rescue StandardError => e
    puts "Unexpected error: #{e.message}"
    puts e.backtrace.first(5)
    exit 1
  end
end

# Run the main function if this script is executed directly
if __FILE__ == $0
  main
end
